File: packages/rn-bluetooth/android/src/main/java/com/rnbluetooth/EventEmitter.kt
/**
 * EventEmitter.kt
 * Handles event emission to JavaScript layer
 */

package com.rnbluetooth

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule

class EventEmitter(private val reactContext: ReactApplicationContext) {

    companion object {
        private const val EVENT_NAME = "RNLCBluetoothEvent"
    }

    /**
     * Send an event to JavaScript
     */
    fun send(event: WritableMap) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(EVENT_NAME, event)
    }

    /**
     * Send device discovered event
     */
    fun sendDeviceDiscovered(
        deviceId: String,
        rssi: Int,
        payload: WritableMap
    ) {
        val event = Arguments.createMap()
        event.putString("type", "deviceDiscovered")
        event.putString("deviceId", deviceId)
        event.putInt("rssi", rssi)
        event.putMap("payload", payload)
        send(event)
    }

    /**
     * Send connection state changed event
     */
    fun sendConnectionStateChanged(
        deviceId: String,
        state: String
    ) {
        val event = Arguments.createMap()
        event.putString("type", "connectionStateChanged")
        event.putString("deviceId", deviceId)
        event.putString("state", state)
        send(event)
    }

    /**
     * Send follow request received event
     */
    fun sendFollowRequestReceived(
        fromDeviceId: String,
        payloadJson: String
    ) {
        val event = Arguments.createMap()
        event.putString("type", "followRequestReceived")
        event.putString("fromDeviceId", fromDeviceId)

        // Parse JSON payload
        try {
            val jsonObject = org.json.JSONObject(payloadJson)
            val payload = Arguments.createMap()

            // Parse the payload structure
            val payloadType = jsonObject.optString("type", "follow-request")
            payload.putString("type", payloadType)

            // Support both "requester" (new) and "follower" (legacy) field names
            val userFieldName = if (jsonObject.has("requester")) "requester" else "follower"
            val userJson = jsonObject.getJSONObject(userFieldName)
            val user = Arguments.createMap()
            user.putString("userId", userJson.getString("userId"))
            user.putString("displayName", userJson.getString("displayName"))
            user.putString("publicKey", userJson.getString("publicKey"))
            if (userJson.has("profilePhoto") && !userJson.isNull("profilePhoto")) {
                user.putString("profilePhoto", userJson.getString("profilePhoto"))
            }

            // Always export as "follower" for backward compatibility with existing event handlers
            payload.putMap("follower", user)

            payload.putString("timestamp", jsonObject.getString("timestamp"))

            event.putMap("payload", payload)
        } catch (e: Exception) {
            // Fallback: send as unparsed string (will likely fail on JS side, but easier to debug)
            android.util.Log.e("EventEmitter", "Failed to parse follow request JSON: ${e.message}")
            android.util.Log.e("EventEmitter", "Payload JSON: $payloadJson")
            event.putString("payloadJson", payloadJson)
        }

        send(event)
    }

    /**
     * Send scan stopped event
     */
    fun sendScanStopped() {
        val event = Arguments.createMap()
        event.putString("type", "scanStopped")
        send(event)
    }

    /**
     * Send error event
     */
    fun sendError(message: String, code: String? = null) {
        val event = Arguments.createMap()
        event.putString("type", "error")
        event.putString("message", message)
        code?.let { event.putString("code", it) }
        send(event)
    }
}

File: packages/rn-bluetooth/ios/EventEmitter.swift
/**
 * EventEmitter.swift
 * Handles event emission to JavaScript layer
 */

import Foundation
import React

@objc(RNLCBluetoothEventEmitter)
class EventEmitter: RCTEventEmitter {

  /// Shared singleton instance
  static var shared: EventEmitter?

  override init() {
    super.init()
    EventEmitter.shared = self
    print("[EventEmitter] 🔌 EventEmitter initialized and set as shared instance")
  }

  @objc override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc override func supportedEvents() -> [String]! {
    return ["RNLCBluetoothEvent"]
  }

  /// Send an event to JavaScript
  /// - Parameter payload: Dictionary containing event data
  func send(_ payload: [String: Any]) {
    sendEvent(withName: "RNLCBluetoothEvent", body: payload)
  }

  /// Send a device discovered event
  func sendDeviceDiscovered(
    deviceId: String,
    rssi: Int,
    payload: [String: Any]
  ) {
    print("[EventEmitter] 📤 Sending deviceDiscovered event: \(deviceId)")
    send([
      "type": "deviceDiscovered",
      "deviceId": deviceId,
      "rssi": rssi,
      "payload": payload
    ])
    print("[EventEmitter] ✅ Event sent to JavaScript")
  }

  /// Send a connection state changed event
  func sendConnectionStateChanged(
    deviceId: String,
    state: String
  ) {
    send([
      "type": "connectionStateChanged",
      "deviceId": deviceId,
      "state": state
    ])
  }

  /// Send a follow request received event
  func sendFollowRequestReceived(
    fromDeviceId: String,
    payloadJson: String
  ) {
    guard let payloadData = payloadJson.data(using: .utf8),
          let payload = try? JSONSerialization.jsonObject(with: payloadData) as? [String: Any] else {
      return
    }

    send([
      "type": "followRequestReceived",
      "fromDeviceId": fromDeviceId,
      "payload": payload
    ])
  }

  /// Send a scan stopped event
  func sendScanStopped() {
    send(["type": "scanStopped"])
  }

  /// Send an error event
  func sendError(message: String, code: String? = nil) {
    var payload: [String: Any] = [
      "type": "error",
      "message": message
    ]
    if let code = code {
      payload["code"] = code
    }
    send(payload)
  }
}
